service/service.go

```diff
@@ -68,6 +72,7 @@ type Service struct {
 	telemetrySettings    component.TelemetrySettings
 	host                 *serviceHost
 	telemetryInitializer *telemetryInitializer
+	collectorConf        *confmap.Conf
 }
 
 func New(ctx context.Context, set Settings, cfg Config) (*Service, error) {
```

Comment:
We are passing in two different representations of the same config, one explicitly as "the config", the other riding along in the "settings".

I think @srikanthccv made a good point [here](https://github.com/open-telemetry/opentelemetry-collector/pull/6833/files#r1246509584) that since these two representations are both passed in through the public API, we are introducing an opportunity for them to be out of sync.

Since we are proposing a breaking change to the way we create a Service, I think we should consider a change that does not introduce this possibility.

I am new to this section of the code so may be overlooking some details but perhaps it would be cleaner to pass the `ConfigProvider` into `service.New`, _instead_ of `service.Config`? This would require moving `otelcol.ConfigProvider` and the `configProvider` implementation into the `service` package, but I think it would be more logically consistent and would remove the need for the new `ConfmapProvider` interface.

Reply:
>This would require moving otelcol.ConfigProvider and the configProvider implementation into the service package, but I think it would be more logically consistent and would remove the need for the new ConfmapProvider interface.

I've spent some time looking into this, and I think an argument could be made for moving these into the `service` package. The `otelcol` package is primarily tasked with setting up the Collector process: OS signal handling, environment variables, command line flag parsing, etc. Output from `otelcol.ConfigProvider` is just passed to the service and isn't used by `otelcol.Collector`, so encapsulating it in the `service` package seems like it would result in a cleaner API. I'll do some additional prototyping to get a sense for whether this is a viable approach.

That said, I have had some difficulties fully understanding the design decisions behind the current API boundaries, so I may be missing something essential. @open-telemetry/collector-maintainers what do you think?

> We are passing in two different representations of the same config, one explicitly as "the config", the other riding along in the "settings".
>
> I think @srikanthccv made a good point [here](https://github.com/open-telemetry/opentelemetry-collector/pull/6833/files#r1246509584) that since these two representations are both passed in through the public API, we are introducing an opportunity for them to be out of sync.

I missed the original point of that comment, but it's a good point, thanks for calling that out. My original goal was to avoid this situation by giving `Config` a method that would convert it to a map-like structure, but that isn't possible without giving it a dependency on `confmap.Conf`. If moving `otelcol.ConfigProvider` to `service` doesn't pan out I'll see if there's a way this can be addressed.

Reply:
> design decisions behind the current API boundaries

I'm was not tuned in when these changes were made but my understanding of the boundary between `otelcol` and `service` is roughly that `otelcol` should be able to run multiple instances of `service.Service` in sequence. I think the changes we are talking about are in line with this intention.

Reply:
`service` module should represent the `service` part of the configuration see https://github.com/open-telemetry/opentelemetry-collector/issues/5564. `ConfigProvider` was there before moving to `otelcol`, see https://github.com/open-telemetry/opentelemetry-collector/pull/6768. Maybe we can move to a new separate module given that it only depends on `confmap` so both otelcol and service can use it, but I cannot find any good place for it.

Reply:
Maybe [`confmap/provider`](https://github.com/open-telemetry/opentelemetry-collector/tree/main/confmap/provider)? There's no code in the package currently but it seems like a natural place.

Reply:
>`service` module should represent the `service` part of the configuration see https://github.com/open-telemetry/opentelemetry-collector/issues/5564. `ConfigProvider` was there before moving to `otelcol`, see https://github.com/open-telemetry/opentelemetry-collector/pull/6768.

Thank you for calling these out. I think most of my confusion around this stems from the fact that it seems like the `service` part of the config is just the whole config. At present, the entirety of the config represented by `otelcol.Config` ultimately only goes to configure the service or one of its subcomponents. Is it expected that it may be possible to configure other aspects of the Collector that aren't in the service? Furthermore, since the service contains all of the Collector's prominent features (reporting self-telemetry, running extensions and pipelines), what would we expect might live outside it?

>Maybe we can move to a new separate module given that it only depends on confmap so both otelcol and service can use it, but I cannot find any good place for it.

>Maybe [confmap/provider](https://github.com/open-telemetry/opentelemetry-collector/tree/main/confmap/provider)? There's no code in the package currently but it seems like a natural place.

I think `ConfigProvider` could possibly make sense in `confmap/provider`, though in that case I would prefer to abstract away the details around the Collector's configuration format since that feels like an orthogonal concern to the `confmap` package.

Reply:
I have a sketch of what moving `ConfigProvider` out of the `otelcol` package would look like here: https://github.com/evan-bradley/opentelemetry-collector/commit/0364f7a031048c681d350e56862967786ff0a441. I've implemented it by moving it to the `service` package, but I think some of the design would still stand.

The assumption I'm making in this is that the distinction between `otelcol` and `service` is that `otelcol` provides the Collector as an application, as where `service` provides it as a library. Depending on the answers to my questions in my previous comment, this way of looking at things may entirely miss the mark.